downloadservice/app.py

```diff
@@ -159,10 +183,94 @@ def login(user):
     return render_template('index.html', user=user, token=token)
 
 
-def get_upstream_session():
+def certificate_validity(certificate):
+    try:
+        x509 = OpenSSL.crypto.load_certificate(
+            OpenSSL.crypto.FILETYPE_PEM, certificate)
+        asn1_time = x509.get_notAfter()
+        return datetime.strptime(asn1_time.decode(), '%Y%m%d%H%M%S%fZ')
+    except Exception:
+        raise CertificateError('invalid certificate')
```

Comment:
would be better to distinguish unparsable, expired, incorrectly signed. at least some of these